tempesta_fw/connection.h

```diff
@@ -284,6 +289,55 @@ tfw_srv_conn_hasnip(TfwSrvConn *srv_conn)
 	return test_bit(TFW_CONN_B_HASNIP, &srv_conn->flags);
 }
 
+/*
+ * Tell if connection is temporary inactive due to full work queue.
+ */
+static inline bool
+tfw_srv_conn_busy(TfwSrvConn *conn)
+{
+	if (test_bit(TFW_CONN_B_BUSY, &conn->flags))
+		return true;
+	rmb();
+	if (time_is_after_jiffies(READ_ONCE(conn->jbusytstamp)))
+		return true;
+
+	return false;
+}
+
+/*
+ * Make connection temporary inactive due to full work queue.
+ */
+static inline void
+tfw_srv_conn_set_busy(TfwSrvConn *conn)
+{
+	if (!test_bit(TFW_CONN_B_BUSY, &conn->flags))
+		set_bit(TFW_CONN_B_BUSY, &conn->flags);
+}
+
+/*
+ * Clear busy bit for server connection (if set).
+ */
+static inline void
+tfw_srv_conn_clear_busy(TfwSrvConn *conn)
+{
+	if (test_bit(TFW_CONN_B_BUSY, &conn->flags))
+		clear_bit(TFW_CONN_B_BUSY, &conn->flags);
```

Comment:
Why test the bit first when the goal is to just clear it? There's nothing here that depends on the bit being set before. Also, this sequence doesn't even provide this guarantee as it's not atomic.

Reply:
It is expected that most of the time Tempesta would work not in overloaded state, i.e. - without `-EBUSY` appearance and without `TFW_CONN_B_BUSY` bit set for connections. So, with this code most of the time - only `test_bit()` call will be executed, which compiled into light read-only instruction:
```
test  imm8, reg/mem8
```
instead of `clear_bit()` call, which in turn compiled into significantly more heavy, locked read-modify-write instruction:
```
lock andb  imm8, mem8
```